### PR TITLE
[InstCombine] Improve select equiv fold for plain condition

### LINKED ISSUE #83225: [InstCombine] Missed optimization: fold select if the user of its select user indicates the condition
Alive2 proof: https://alive2.llvm.org/ce/z/EDjfon

### Motivating example 

```llvm
define i32 @src(i32 %a, i1 %c1, i1 %c2){
entry:
  %s1 = select i1 %c1, i32 23, i32 45
  %s2 = select i1 %c2, i32 666, i32 %s1
  %s3 = select i1 %c1, i32 789, i32 %s2
  ret i32 %s3
}
```
can be folded to:
```llvm
define i32 @tgt(i32 %a, i1 %c1, i1 %c2){
entry:
  %s2 = select i1 %c2, i32 666, i32 45
  %s3 = select i1 %c1, i32 789, i32 %s2
  ret i32 %s3
}
```

### Real-world motivation

This snippet of IR is derived from [libdeflate/lib/crc32.c@dispatch_crc32](https://github.com/ebiggers/libdeflate/blob/e93127b7e85d8dc65a80ed0ad4883c4b3efbe813/lib/crc32.c#L241) (after O3 pipeline, original IR comes from [llvm-opt-benchmark](https://github.com/dtcxzyw/llvm-opt-benchmark)).
The example above is a reduced version. If you're interested in the original suboptimal IR and optimal IR, see also:https://godbolt.org/z/rjf9zWM81

**Let me know if you can confirm that it's an optimization opportunity, thanks.**

### PR DESCRIPTION
**Pull request by vfdff**
Fold select if the user of its select user indicates the condition

Fix https://github.com/llvm/llvm-project/issues/83225

### CODE REVIEW

File: llvm/test/Transforms/InstCombine/select.ll [hunk 1]
```diff
@@ -3672,3 +3672,33 @@ define i32 @src_select_xxory_eq0_xorxy_y(i32 %x, i32 %y) {
   %cond = select i1 %xor0, i32 %xor, i32 %y
   ret i32 %cond
 }
+
+define i32 @sequence_select_with_same_cond (i1 %c1, i1 %c2){
```
[the remark this answers is not in the document]
Comment (vfdff) [reply llvm/test/Transforms/InstCombine/select.ll:3676]: oh, yes, it is strange, it works when I only keep this case **@sequence_select_with_same_cond**

[the remark this answers is not in the document]
Comment (vfdff) [reply llvm/test/Transforms/InstCombine/select.ll:3676]: > oh, yes, it is strange, it works when I only keep this case **@sequence_select_with_same_cond**

with a few debug, I find that the const value "i32 45" in different function will be seen as the same variable, so it doesn't match the **m_OneUse** of the m_Value(FalseNext) for **@sequence_select_with_same_cond** when I add a extra test sequence_select_with_same_cond_extra_use.

**Comment by vfdff** [comment]
> I think this should be a minor generalization of some existing simplifyWithOpReplaced or foldSelectValueEquivalence style fold.
Thanks, refactor with simplifyWithOpReplaced. （foldSelectValueEquivalence is used for **select + cmp**）

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp [hunk 2]
```diff
@@ -500,6 +500,42 @@ static bool isSelect01(const APInt &C1I, const APInt &C2I) {
...
+  Value *FalseVal = SINext->getFalseValue();
+  Value *CondNext;
+  Value *FalseNext;
+  while (match(FalseVal,
+               m_Select(m_Value(CondNext), m_Value(), m_Value(FalseNext)))) {
+    // If the type of select is not an integer type or if the condition and
+    // the selection type are not both scalar nor both vector types, there is no
+    // point in attempting to match these patterns.
+    if (!isa<Constant>(CondVal) && SelType->isIntOrIntVectorTy() &&
+        CondType->isVectorTy() == SelType->isVectorTy())
```
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:522]: Instead of all this why not just match `m_Select(m_Specific(CondVal), ...)`?

Also, why are you only chasing false val?

Comment (vfdff) [reply llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:522]: Thanks, also chase true value now.
I don't use `m_Select(m_Specific(CondVal), ...)` because we don't need all the select chains have the same condition.
Just as the case **@src** in [#83225](https://github.com/llvm/llvm-project/issues/83225) shows, the condition of **%s2** is different from that of **%s3**.
```
define i32 @src(i32 %a, i1 %c1, i1 %c2){
entry:
  %s1 = select i1 %c1, i32 23, i32 45
  %s2 = select i1 %c2, i32 666, i32 %s1
  %s3 = select i1 %c1, i32 789, i32 %s2
  ret i32 %s3
}
```

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp [hunk 3]
```diff
@@ -500,6 +500,51 @@ static bool isSelect01(const APInt &C1I, const APInt &C2I) {
...
+      if (Value *S = simplifyWithOpReplaced(TrueVal, CondVal,
+                                            ConstantInt::getTrue(CondType), SQ,
+                                            /* AllowRefinement */ true)) {
+        IC.replaceOperand(*SINext, 1, S);
+        return true;
+      }
+
+    SINext = cast<SelectInst>(TrueVal);
+    TrueVal = SINext->getTrueValue();
+  }
```
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:543]: Can you put the codes in a lambda, think only argument you need is `OpNumber`.

Comment (vfdff) [reply llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp]: Apply your comment, thanks @goldsteinn

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp [hunk 4]
```diff
@@ -500,6 +500,48 @@ static bool isSelect01(const APInt &C1I, const APInt &C2I) {
...
+        }
+
+      SINext = cast<SelectInst>(ValOp);
+      ValOp = SINext->getOperand(OpIndex);
+    }
+    return false;
+  };
+
+  // Try to simplify the true value of select.
+  if (TrysimplifySeqSelect(1))
```
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:535]: `/*OpIndx=*/1`

likewise below.

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp [hunk 5]
```diff
@@ -500,6 +500,48 @@ static bool isSelect01(const APInt &C1I, const APInt &C2I) {
...
+                                          const SimplifyQuery &SQ,
+                                          InstCombinerImpl &IC) {
+  Value *CondVal = SI.getCondition();
+  if (isa<Constant>(CondVal))
+    return false;
+
+  auto TrysimplifySeqSelect = [=, &SI, &IC](unsigned OpIndex) {
+    assert((OpIndex == 1 || OpIndex == 2) && "Unexpected operand index");
+    Type *CondType = CondVal->getType();
+    Value *RepOp = OpIndex == 1 ? ConstantInt::getTrue(CondType)
```
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:515]: RepOp = SI.getOperand(OpIndex);

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp [hunk 6]
```diff
@@ -500,6 +500,48 @@ static bool isSelect01(const APInt &C1I, const APInt &C2I) {
...
+                                : ConstantInt::getFalse(CondType);
+    SelectInst *SINext = &SI;
+    Value *ValOp = SINext->getOperand(OpIndex);
+    Value *CondNext;
+    while (match(ValOp, m_Select(m_Value(CondNext), m_Value(), m_Value()))) {
+      if (CondNext == CondVal)
+        if (Value *S = simplifyWithOpReplaced(ValOp, CondVal, RepOp, SQ,
```
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:522]: Think this is invalid if `ValOp` has multiple uses as you are simplifying based on the context of a single use.

Think you can just do `IC.replaceOperand(*SINext, OpIndx, cast<SelectInst>(ValOp)->getOperand(OpIndex)`.

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp [hunk 7]
```diff
@@ -500,6 +500,48 @@ static bool isSelect01(const APInt &C1I, const APInt &C2I) {
...
+  auto TrysimplifySeqSelect = [=, &SI, &IC](unsigned OpIndex) {
+    assert((OpIndex == 1 || OpIndex == 2) && "Unexpected operand index");
+    Type *CondType = CondVal->getType();
+    Value *RepOp = OpIndex == 1 ? ConstantInt::getTrue(CondType)
+                                : ConstantInt::getFalse(CondType);
+    SelectInst *SINext = &SI;
+    Value *ValOp = SINext->getOperand(OpIndex);
+    Value *CondNext;
+    while (match(ValOp, m_Select(m_Value(CondNext), m_Value(), m_Value()))) {
+      if (CondNext == CondVal)
```
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:521]: You could also handle `match(CondNext, m_Not(CondVal))` and flip OpIndex for the replacement.

Comment (vfdff) [reply llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:521]: Thanks, I'll address that with a following seperate patch

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp (hunk 6 above)
Comment (vfdff) [reply llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp]: Apply your comment, thanks

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp (hunk 4 above)
Comment (vfdff) [reply llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp]: Done, thanks

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp [hunk 8]
```diff
@@ -519,9 +519,13 @@ static bool simplifySeqSelectWithSameCond(SelectInst &SI,
     Value *CondNext;
     while (match(ValOp, m_Select(m_Value(CondNext), m_Value(), m_Value()))) {
       if (CondNext == CondVal)
-        if (Value *S = simplifyWithOpReplaced(ValOp, CondVal, RepOp, SQ,
-                                              /* AllowRefinement */ true)) {
-          IC.replaceOperand(*SINext, OpIndex, S);
+        if (simplifyWithOpReplaced(ValOp, CondVal, RepOp, SQ,
+                                   /* AllowRefinement */ true)) {
```
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:523]:
```suggestion
                                   /*AllowRefinement=*/ true)) {
```

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp [hunk 9]
```diff
@@ -500,6 +500,52 @@ static bool isSelect01(const APInt &C1I, const APInt &C2I) {
...
+    return false;
+  };
+
+  // Try to simplify the true value of select.
+  if (TrysimplifySeqSelect(/*OpIndx=*/1))
+    return true;
+
+  // Try to simplify the false value of select.
+  if (TrysimplifySeqSelect(/*OpIndx=*/2))
+    return true;
```
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:544]: nit: misspelled `OpIndex`

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp [hunk 10]
```diff
@@ -500,6 +500,52 @@ static bool isSelect01(const APInt &C1I, const APInt &C2I) {
...
+    Type *CondType = CondVal->getType();
+    Value *RepOp = OpIndex == 1 ? ConstantInt::getTrue(CondType)
+                                : ConstantInt::getFalse(CondType);
+    SelectInst *SINext = &SI;
+    Value *ValOp = SINext->getOperand(OpIndex);
+    Value *CondNext;
+    while (match(ValOp, m_Select(m_Value(CondNext), m_Value(), m_Value()))) {
+      if (CondNext == CondVal)
+        if (simplifyWithOpReplaced(ValOp, CondVal, RepOp, SQ,
+                                   /* AllowRefinement */ true)) {
```
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:523]: I don't understand what the simplifyWithOpReplaced is doing anymore.

**Comment by nikic** [comment]
FWIW, what I had in mind with my comment was more something along the lines of https://github.com/llvm/llvm-project/pull/85663, which generalizes the existing select equivalence fold. The disadvantage is that it does not handle the multi-use example out of the box (but could be generalized).

**Comment by vfdff** [comment]
> FWIW, what I had in mind with my comment was more something along the lines of #85663, which generalizes the existing select equivalence fold. The disadvantage is that it does not handle the multi-use example out of the box (but could be generalized).

hi @nikic, I saw #85663 is closed because the compile time regression, so would you suggest I continue with current solution?

**Comment by arsenm** [comment]
Title isn't clear to me

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp (hunk 10 above)
Comment (vfdff) [reply llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp]: deleted the redundant call `simplifyWithOpReplaced `, thanks

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp (hunk 9 above)
Comment (vfdff) [reply llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:537]: Done

**Comment by dtcxzyw** [comment]
@vfdff Could you please rebase this patch on main?

**Comment by vfdff** [comment]
> @vfdff Could you please rebase this patch on main?

Thanks for the reminder. I found that the test cases have been committed in advance. (**commit bd9a2afa**)

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp [hunk 11]
```diff
@@ -500,6 +500,45 @@ static bool isSelect01(const APInt &C1I, const APInt &C2I) {
   return C1I.isOne() || C1I.isAllOnes() || C2I.isOne() || C2I.isAllOnes();
 }
 
+/// Try to simplify a select instruction when the user of its select user
+/// indicates the condition.
+static bool simplifySeqSelectWithSameCond(SelectInst &SI,
+                                          const SimplifyQuery &SQ,
+                                          InstCombinerImpl &IC) {
+  Value *CondVal = SI.getCondition();
+  if (isa<Constant>(CondVal))
```
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:509]: match(CondVal, m_ImmConstant())

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp [hunk 12]
```diff
@@ -500,6 +500,45 @@ static bool isSelect01(const APInt &C1I, const APInt &C2I) {
...
+    while (match(ValOp, m_Select(m_Value(CondNext), m_Value(), m_Value()))) {
+      if (CondNext == CondVal) {
+        IC.replaceOperand(*SINext, OpIndex,
+                          cast<SelectInst>(ValOp)->getOperand(OpIndex));
+        return true;
+      }
+
+      SINext = cast<SelectInst>(ValOp);
+      ValOp = SINext->getOperand(OpIndex);
+    }
```
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:527]: Im not sure concerned about this, I think it may be slight better to continue looping until no more select and have a changed boolean.
Take a case like:
```
define i8 @foo(i1 %cond0, i1 %cond1, i8 %w, i8 %x, i8 %y, i8 %z, i8 %a) {
  %sel0 = select i1 %cond0, i8 %w, i8 %x
  %sel1 = select i1 %cond1, i8 %y, i8 %sel0
  %sel2 = select i1 %cond0, i8 %sel1, i8 %z
  %sel3 = select i1 %cond1, i8 %sel2, i8 %a
  ret i8 %sel3
}
```

I think what happens here is you replace `%sel2 = select i1 %cond0, i8 %sel1, i8 %z` -> `%sel2 = select i1 %cond0, i8 %y, i8 %z`. But after that you won't be able to replace ` %sel1 = select i1 %cond1, i8 %y, i8 %sel0` -> ` %sel1 = select i1 %cond1, i8 %y, i8 %x` b.c you won't match any `select` arm in `%sel2`.

I'm not really sure if this ever comes up, I don't think its a must fix, just a thought.

Comment (vfdff) [reply llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:527]: Yes, I think it need anther patch to handle **match(CondNext, m_Not(CondVal))** and flip OpIndex for the replacement as your previous commit, https://alive2.llvm.org/ce/z/yGCRqB.

**Comment by goldsteinn** [comment]
LGTM.

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp (hunk 11 above)
Comment (vfdff) [reply llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp]: done, thanks

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp (hunk 12 above)
Comment (vfdff) [reply llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:527]: Thansk, need a seperate patch to improve the case  https://alive2.llvm.org/ce/z/YGZPFa

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp [hunk 13]
```diff
@@ -500,6 +500,45 @@ static bool isSelect01(const APInt &C1I, const APInt &C2I) {
...
+    return false;
+  };
+
+  // Try to simplify the true value of select.
+  if (TrysimplifySeqSelect(/*OpIndex=*/1))
```
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:532]: Capitalization weird. trySimplify...

Comment (vfdff) [reply llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp]: Fixed, thanks

**Comment by vfdff** [comment]
ping ？

**Comment by goldsteinn** [comment]
Still LGTM, ping @dtcxzyw @nikic

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp [hunk 14]
```diff
@@ -500,6 +500,44 @@ static bool isSelect01(const APInt &C1I, const APInt &C2I) {
...
 }
 
+/// Try to simplify a select instruction when the user of its select user
+/// indicates the condition.
+static bool simplifySeqSelectWithSameCond(SelectInst &SI,
+                                          const SimplifyQuery &SQ,
+                                          InstCombinerImpl &IC) {
+  Value *CondVal = SI.getCondition();
+  if (match(CondVal, m_ImmConstant()))
+    return false;
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:510]: Isn't it handled by `simplifySelectInst`?

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp [hunk 15]
```diff
@@ -500,6 +500,44 @@ static bool isSelect01(const APInt &C1I, const APInt &C2I) {
...
+
+  auto trySimplifySeqSelect = [=, &SI, &IC](unsigned OpIndex) {
+    assert((OpIndex == 1 || OpIndex == 2) && "Unexpected operand index");
+    SelectInst *SINext = &SI;
+    Value *ValOp = SINext->getOperand(OpIndex);
+    Value *CondNext;
+    while (match(ValOp, m_Select(m_Value(CondNext), m_Value(), m_Value()))) {
+      if (CondNext == CondVal) {
+        IC.replaceOperand(*SINext, OpIndex,
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:519]: It is incorrect when the inner select is used by both arms.

Example: https://alive2.llvm.org/ce/z/sv6m69
```
define i8 @src(i1 %cond0, i1 %cond1, i8 %a, i8 %b) {
entry:
%sel0 = select i1 %cond0, i8 %a, i8 %b
%sel1 = select i1 %cond1, i8 %sel0, i8 %b
%sel2 = select i1 %cond1, i8 %sel1, i8 %b
%sel3 = select i1 %cond0, i8 %sel1, i8 %sel2
ret i8 %sel3
}

define i8 @tgt(i1 %cond0, i1 %cond1, i8 %a, i8 %b) {
entry:
%sel0 = select i1 %cond0, i8 %a, i8 %b
%sel1 = select i1 %cond1, i8 %a, i8 %b
%sel2 = select i1 %cond1, i8 %sel1, i8 %b
%sel3 = select i1 %cond0, i8 %sel1, i8 %sel2
ret i8 %sel3
}
```

**Review by dtcxzyw — CHANGES_REQUESTED** [review]

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp (hunk 14 above)
Comment (vfdff) [reply llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:510]: Yes, it isn't.
[#85663](https://github.com/llvm/llvm-project/pull/85663) is a more general improvement by nikic, and it has some compile time regression, so I continue on this solution.

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp (hunk 15 above)
Comment (vfdff) [reply llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:519]: hi @dtcxzyw , the above case can be addressed correctly https://alive2.llvm.org/ce/z/HyqtF4, and it will replace the **%sel2** because we find the first 3 selects  match the combine pattern
 - note:  here it doesn't update the %sel1, but updates the %sel2
```
define i8 @tgt(i1 %cond0, i1 %cond1, i8 %a, i8 %b) {
entry:
%sel0 = select i1 %cond0, i8 %a, i8 %b
%sel1 = select i1 %cond1, i8 %sel0, i8 %b
%sel2 = select i1 %cond1, i8 %sel0, i8 %b  ;   %sel1 --> %sel0
%sel3 = select i1 %cond0, i8 %sel1, i8 %sel2
ret i8 %sel3
}
```

Comment (vfdff) [reply llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp]: > Think you can get around that by just building a new select.
Thanks @goldsteinn , apply your comment to make the PR more robust.

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp (hunk 14 above)
Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:511]: https://github.com/llvm/llvm-project/blob/a61252419779a6d4a5ebf71e7e2fc4adc75cfddd/llvm/lib/Analysis/InstructionSimplify.cpp#L4767-L4771

Can you provide a test to show that the early exit here is necessary?

Comment (vfdff) [reply llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:511]: Thanks, I delete the early exit checking.
I added it just because I thought it was an optimized expression and didn't need to be optimized.

File: llvm/test/Transforms/InstCombine/select.ll [hunk 16]
```diff
@@ -3759,3 +3759,18 @@ define i32 @sequence_select_with_same_cond_extra_use(i1 %c1, i1 %c2){
...
+
+define i8 @sequence_select_with_same_cond_multi_arms(i1 %cond0, i1 %cond1, i8 %a, i8 %b) {
+; CHECK-LABEL: @sequence_select_with_same_cond_multi_arms(
+; CHECK-NEXT:    [[SEL2:%.*]] = select i1 [[TMP1:%.*]], i8 [[A:%.*]], i8 [[B:%.*]]
+; CHECK-NEXT:    [[SEL1_NEW:%.*]] = select i1 [[COND1:%.*]], i8 [[A]], i8 2
+; CHECK-NEXT:    [[SEL4:%.*]] = select i1 [[COND1]], i8 [[SEL2]], i8 3
+; CHECK-NEXT:    [[SEL3:%.*]] = select i1 [[TMP1]], i8 [[SEL1_NEW]], i8 [[SEL4]]
+; CHECK-NEXT:    ret i8 [[SEL3]]
+;
+  %sel0 = select i1 %cond0, i8 %a, i8 %b
```
Comment (vfdff) [inline llvm/test/Transforms/InstCombine/select.ll:3771]: * this case  has 2  transformation in combine, https://alive2.llvm.org/ce/z/ChFi5F
  - first
  ```
define i8 @fist(i1 %cond0, i1 %cond1, i8 %a, i8 %b) {
  %sel0 = select i1 %cond0, i8 %a, i8 %b
  %sel1 = select i1 %cond1, i8 %sel0, i8 2
  %sel2 = select i1 %cond1, i8 %sel0, i8 3 ; update directly base on first 3 nodes
  %sel3 = select i1 %cond0, i8 %sel1, i8 %sel2
  ret i8 %sel3
}
  ```
  - second
  ```
define i8 @second(i1 %cond0, i1 %cond1, i8 %a, i8 %b) {
    %sel0 = select i1 %cond0, i8 %a, i8 %b
    %sel1.new = select i1 %cond1, i8 %a, i8 2
    %sel2 = select i1 %cond1, i8 %sel0, i8 3
    %sel3 = select i1 %cond0, i8 %sel1.new, i8 %sel2; base on the 4 nodes
    ret i8 %sel3
  }
  ```

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp [hunk 17]
```diff
@@ -500,6 +500,65 @@ static bool isSelect01(const APInt &C1I, const APInt &C2I) {
   return C1I.isOne() || C1I.isAllOnes() || C2I.isOne() || C2I.isAllOnes();
 }
 
+/// Try to simplify a select instruction when the user of its select user
+/// indicates the condition.
```
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:504]: I don't know how to parse this comment "when the user of its select user". Can you rephrase this whole thing?

File: llvm/test/Transforms/InstCombine/select.ll [hunk 18]
```diff
@@ -3762,3 +3759,18 @@ define i32 @sequence_select_with_same_cond_extra_use(i1 %c1, i1 %c2){
...
+  %sel1 = select i1 %cond1, i8 %sel0, i8 2; %sel1 used in multi arms
+  %sel2 = select i1 %cond1, i8 %sel1, i8 3
+  %sel3 = select i1 %cond0, i8 %sel1, i8 %sel2
+  ret i8 %sel3
+}
```
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/select.ll:3776]: Can you add some tests that show flag preservation behavior?

Comment (vfdff) [reply llvm/test/Transforms/InstCombine/select.ll:3801]: Added `sequence_select_with_same_cond_float_and_fmf_flag1` and `sequence_select_with_same_cond_float_and_fmf_flag2`, thanks

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp (hunk 17 above)
Comment (vfdff) [reply llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp]: Add comment with a example.

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp (hunk 15 above)
Comment (vfdff) [reply llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp]: I think we don't need support the case with multi-arms because it may be not profitable  to build a new select
```
define i32 @sequence_select_with_same_cond_extra_use(i1 %c1, i1 %c2){
; CHECK-LABEL: @sequence_select_with_same_cond_extra_use(
; CHECK-NEXT:    [[S1:%.*]] = select i1 [[C1:%.*]], i32 23, i32 45
; CHECK-NEXT:    [[S2_NEW:%.*]] = select i1 [[C2:%.*]], i32 666, i32 45
; CHECK-NEXT:    [[S2:%.*]] = select i1 [[C2]], i32 666, i32 [[S1]]
; CHECK-NEXT:    call void @use32(i32 [[S2]])
; CHECK-NEXT:    [[S3:%.*]] = select i1 [[C1]], i32 789, i32 [[S2_NEW]]
; CHECK-NEXT:    ret i32 [[S3]]
;
  %s1 = select i1 %c1, i32 23, i32 45
  %s2 = select i1 %c2, i32 666, i32 %s1
  call void @use32(i32 %s2)
  %s3 = select i1 %c1, i32 789, i32 %s2
  ret i32 %s3
}
```

**Comment by vfdff** [comment]
According [performace considerations](https://github.com/llvm/llvm-project/pull/83405#discussion_r1598042801), I deleted the support with multi-use select node and rebase to top upstream as the code conflicts.

**Comment by vfdff** [comment]
ping @dtcxzyw, @arsenm

**Comment by vfdff** [comment]
close to avoid blocking other solutions

**Comment by arsenm** [comment]
> close to avoid blocking other solutions

What do you mean? If there's an alternative patch, it should be mentioned in the closing message. Otherwise, why close it

**Comment by vfdff** [comment]
Oh，sorry,  I reopen it now.
I close it because the solution is limit to `single-use select node`, which seems not perfect :)

**Comment by arsenm** [comment]
> Oh，sorry, I reopen it now. I close it because the solution is limit to `single-use select node`, which seems not perfect :)

That is the solution to many combine problems. It's a common and reasonable restriction

File: llvm/test/Transforms/InstCombine/select.ll (hunk 18 above)
Comment (arsenm) [reply llvm/test/Transforms/InstCombine/select.ll:4653]: Better to have tests with different flags (and more specific than fast). Need to see the union / intersect behavior

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp [hunk 19]
```diff
@@ -500,6 +500,51 @@ static bool isSelect01(const APInt &C1I, const APInt &C2I) {
...
+    assert((OpIndex == 1 || OpIndex == 2) && "Unexpected operand index");
+    SelectInst *SINext = &SI;
+    Type *SelType = SINext->getType();
+    Value *ValOp = SINext->getOperand(OpIndex);
+    Value *CondNext;
+    // Don't need propagate FMF flag because we update the operand of SINext
+    // directly.
+    // It is not profitable to build a new select for SINext with multi-arms.
+    while (match(ValOp, m_Select(m_Value(CondNext), m_Value(), m_Value()))) {
+      if (CondNext == CondVal && SINext->hasOneUse()) {
```
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:524]: It should be `match(ValOp, m_OneUse(m_Select(....))` then drop the `SINext->hasOneUse()`, otherwise this won't fold the base select if it has multiple uses.

Comment (vfdff) [reply llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:524]: For the following case, it can be optimized because **%sel1** is a sinle use node.
If we adjust according above change, then it will not optimize because the operand of **%sel1** （%sel0）is multi used.
```
define i8 @sequence_select_with_same_cond_multi_arms_src(i1 %cond0, i1 %cond1, i8 %a, i8 %b) {
  %sel0 = select i1 %cond0, i8 %a, i8 %b
  %sel1 = select i1 %cond1, i8 %sel0, i8 2; %sel1 used in single node
  %sel2 = select i1 %cond1, i8 %sel0, i8 3
  %sel3 = select i1 %cond0, i8 %sel1, i8 %sel2
  ret i8 %sel3
}

define i8 @sequence_select_with_same_cond_multi_arms_tgt(i1 %cond0, i1 %cond1, i8 %a, i8 %b) {
  %sel0 = select i1 %cond0, i8 %a, i8 %b
  %sel1 = select i1 %cond1, i8 %a, i8 2; %sel1 used in single node
  %sel2 = select i1 %cond1, i8 %sel0, i8 3
  %sel3 = select i1 %cond0, i8 %sel1, i8 %sel2
  ret i8 %sel3
}
```

Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:524]: Makes sense, what about `V == SINext || SINext->hasOneUse()` then?

Comment (vfdff) [reply llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:524]: do you mean **ValOp == SINext || SINext->hasOneUse()** ? the condition **ValOp == SINext** is protected by **ValOp = SINext->getOperand(OpIndex)**, and the **CondNext == CondVal** is also need, consider case. https://alive2.llvm.org/ce/z/TKLmMG
```
define i32 @src(i32 %a, i1 %c1, i1 %c2, i1 %c3){
  %s1 = select i1 %c1, i32 23, i32 45
  %s2 = select i1 %c2, i32 666, i32 %s1 ; this node can be optimized iff %c1 == %c3
  %s3 = select i1 %c3, i32 789, i32 %s2
  ret i32 %s3
}

define i32 @tgt(i32 %a, i1 %c1, i1 %c2, i1 %c3){
  %s1 = select i1 %c1, i32 23, i32 45
  %s2 = select i1 %c2, i32 666, i32 45
  %s3 = select i1 %c3, i32 789, i32 %s2
  ret i32 %s3
}
```

**Comment by dtcxzyw** [comment]
Ping. I need this patch to unblock https://github.com/llvm/llvm-project/pull/119884.
I noticed this patch when drafting an alternative with `replaceInInstruction`.

**Comment by vfdff** [comment]
hi @dtcxzyw I don't work on this issue now, please update your MR at your convenience
